### PR TITLE
Support ordinal bin.

### PR DESCRIPTION
**Pull request by domoritz**
Before:
screen shot 2017-08-12 at 23 44 07
screen shot 2017-08-13 at 11 29 55

After:
screen shot 2017-08-13 at 11 27 56
screen shot 2017-08-12 at 23 53 51
screen shot 2017-08-13 at 11 28 50

Reverts https://github.com/vega/vega-lite/issues/2773

### CODE REVIEW

**Comment by domoritz** [comment]
Let me try to tweak the second example a bit.

**Comment by domoritz** [comment]
Ha, and now

screen shot 2017-08-13 at 11 47 43

File: src/compile/data/bin.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ function rangeFormula(model: ModelWithField, fieldDef: FieldDef<string>, channel
 
       return {
         formulaAs: field(fieldDef, {binSuffix: 'range'}),
-        formula: `${numberFormatExpr(startField, guide.format, config)} + " - " + ${numberFormatExpr(endField, guide.format, config)}`
+        formula: `${startField} === null ? 'null' : ${numberFormatExpr(startField, guide.format, config)} + " - " + ${numberFormatExpr(endField, guide.format, config)}`
```
Comment (kanitw) [inline src/compile/data/bin.ts]: What if NaN and undefined?   Better use the same expression we use for filtering invalid value?

**Comment by kanitw** [comment]
I don't understand why [this commit](https://github.com/vega/vega-lite/pull/2864/commits/ebde6e7efe10f48cf2840dac5c6d6d2b50d1c5a6) help 
"support ordinal binned facet."  -- Updating test (and examples, if it is not too slow) within the commits rather than having "fix tests and update examples" commits at the end for the whole PR would help better.

File: src/compile/axis/rules.ts [hunk 2]
```diff
@@ -78,11 +79,10 @@ export function orient(channel: Channel) {
...
 export function tickCount(channel: Channel, fieldDef: FieldDef<string>, scaleType: ScaleType, size: VgSignalRef) {
 
-  if (!fieldDef.bin && !hasDiscreteDomain(scaleType) && scaleType !== 'log') {
-    // Vega's default tickCount often lead to a lot of label occlusion on X without 90 degree rotation
-    // Thus, we set it to 5 for width = 200
-    // and set the same value for y for consistency.
-
+  if (!hasDiscreteDomain(scaleType) && scaleType !== 'log') {
+    if (fieldDef.bin) {
+      return {signal: `min(ceil(${size.signal}/40), ${(fieldDef.bin as Bin).maxbins})`};
```
Comment (kanitw) [inline src/compile/axis/rules.ts]: Why removing comments?  Instead, please add comments saying why it should be min of the two?

**Comment by kanitw** [comment]
> I don't understand why [this commit](https://github.com/vega/vega-lite/pull/2864/commits/ebde6e7efe10f48cf2840dac5c6d6d2b50d1c5a6) help 
"support ordinal binned facet."  -- Updating test (and examples, if it is not too slow) within the commits rather than having "fix tests and update examples" commits at the end for the whole PR would help better.  


NVM.  I think I get it now.

File: src/compile/data/bin.ts (hunk 1 above)
Comment (domoritz) [inline src/compile/data/bin.ts]: Done

File: src/compile/axis/rules.ts (hunk 2 above)
Comment (domoritz) [inline src/compile/axis/rules.ts]: Done.

**Comment by kanitw** [comment]
1) Note that this PR introduces two things:
- Support binning with discrete position scales (band/point)
- Support binning with `type: ordinal`.

For the latter, I remember we discuss whether type should mean type before or after transforming, but I can't remember what's our conclusive agreement.  In this case, it would be "type after transforming".  I guess that's fine since `aggregate: "distinct"` requires `type: "quantitatitive"` too.  
I wonder if we have any case where type is type before transforming though.  If not, we are in good shape.  

2)Another question -- This PR seems to now allow `type: nominal` for bin.  Should we prevent/allow that? (or at least throw warning.

**Comment by domoritz** [comment]
> Another question -- This PR seems to now allow type: nominal for bin. Should we prevent/allow that? (or at least throw warning.

I think nominal is totally fine to do. Updated the pr. 

screen shot 2017-08-13 at 14 02 31

**Comment by domoritz** [comment]
idk, I might have a vega bug here but will need to investigate

```
{
      "scale": "y",
      "labelOverlap": true,
      "orient": "left",
      "tickCount": {"signal": "ceil(height/40)"},
      "title": "Number of Records",
      "zindex": 1
    },
...
 "legends": [
    {
      "fill": "color",
      "title": "BIN(IMDB_Rating)",
      "encode": {
        "symbols": {"update": {"shape": {"value": "square"}}}
      }
    }
  ],
```

**Comment by domoritz** [comment]
Confirmed as vega bug (or we are using it wrong)

screen shot 2017-08-13 at 14 38 15

**Pull request merged**